From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Praschl <praschlpaul@g-p.at>
Date: Tue, 4 Aug 2026 23:29:46 +0200
Subject: [PATCH 108/117] Input: apple_z2: Fix D11 runtime result reads

D11 returns its Gen2 runtime result through a transfer that is at
least 64 bytes long, even when the packet header advertises a shorter
logical packet.

Keep validation and dispatch limited to pkt_len while extending only
the SPI wire transfer.

Signed-off-by: Paul Praschl <praschlpaul@g-p.at>

Orchard-replay-of: 7800da9b7edacccf8bafa9486d01cffe9778d1dc
---
 drivers/input/touchscreen/apple_z2.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/input/touchscreen/apple_z2.c b/drivers/input/touchscreen/apple_z2.c
index 135857070..6e4aad4a2 100644
--- a/drivers/input/touchscreen/apple_z2.c
+++ b/drivers/input/touchscreen/apple_z2.c
@@ -49,6 +49,7 @@
 #define APPLE_Z2_CMD_WAKE                0xEE
 #define APPLE_Z2_CMD_SIZE                16
 #define APPLE_Z2_RAW_XFER_MAX_SIZE       64
+#define APPLE_Z2_GEN2_MIN_RESULT_SIZE    64
 #define APPLE_Z2_HBPP_CMD_BLOB           0x3001
 #define APPLE_Z2_FW_MAGIC                0x5746325A
 #define APPLE_Z2_RX_BUF_SIZE             4000
@@ -535,6 +536,7 @@ static int apple_z2_read_packet(struct apple_z2 *z2)
 	struct spi_transfer xfer = { };
 	int error;
 	size_t pkt_len;
+	size_t wire_len;
 	u16 payload_len;
 	u8 counter;
 
@@ -562,12 +564,16 @@ static int apple_z2_read_packet(struct apple_z2 *z2)
 		dev_warn(&z2->spidev->dev, "packet too large: %zu\n", pkt_len);
 		return -EMSGSIZE;
 	}
+	wire_len = pkt_len;
+	if (apple_z2_is_d11(z2))
+		wire_len = max_t(size_t, wire_len,
+				 APPLE_Z2_GEN2_MIN_RESULT_SIZE);
 
 	if (apple_z2_is_j172(z2) || apple_z2_is_iphone7_plus(z2)) {
-		memset(z2->rx_buf, 0xa5, pkt_len);
+		memset(z2->rx_buf, 0xa5, wire_len);
 		xfer.tx_buf = z2->rx_buf;
 		xfer.rx_buf = z2->rx_buf;
-		xfer.len = pkt_len;
+		xfer.len = wire_len;
 		error = spi_sync_transfer(z2->spidev, &xfer, 1);
 	} else {
 		error = spi_read(z2->spidev, z2->rx_buf, pkt_len);
